x/twap/logic_test.go

```diff
@@ -42,76 +43,287 @@ func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
+				s.Require().True(newRecord.P0ArithmeticTwapAccumulator.GT(zeroDec))
+				s.Require().True(newRecord.P1ArithmeticTwapAccumulator.GT(zeroDec))
+			} else {
+				s.Require().False(newRecord.P0ArithmeticTwapAccumulator.GT(zeroDec))
+				s.Require().False(newRecord.P1ArithmeticTwapAccumulator.GT(zeroDec))
+			}
+		})
+	}
+}
+
+// TestPruneRecords tests that all twap records earlier than
+// current block time - RecordHistoryKeepPeriod are pruned from the store.
+func (s *TestSuite) TestPruneRecords() {
+	recordHistoryKeepPeriod := s.twapkeeper.RecordHistoryKeepPeriod(s.Ctx)
+
+	tMin2Record, tMin1Record, baseRecord, tPlus1Record := s.createTestRecordsFromTime(baseTime.Add(-recordHistoryKeepPeriod))
+
+	// non-ascending insertion order.
+	recordsToPreSet := []types.TwapRecord{tPlus1Record, tMin1Record, baseRecord, tMin2Record}
+
+	expectedKeptRecords := []types.TwapRecord{baseRecord, tPlus1Record}
+	s.SetupTest()
+	s.preSetRecords(recordsToPreSet)
+
+	ctx := s.Ctx
+	twapKeeper := s.twapkeeper
+
+	ctx = ctx.WithBlockTime(baseTime)
+
+	err := twapKeeper.PruneRecords(ctx)
+	s.Require().NoError(err)
+
+	s.validateExpectedRecords(expectedKeptRecords)
+}
+
+func TestRecordWithUpdatedAccumulators(t *testing.T) {
+	poolId := uint64(1)
+	tests := map[string]struct {
+		record    types.TwapRecord
+		newTime   time.Time
+		expRecord types.TwapRecord
+	}{
+		"accum with zero value": {
+			record:    newRecord(poolId, time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
+			newTime:   time.Unix(2, 0),
+			expRecord: newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
 		"small starting accumulators": {
-			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
-			interpolateTime: time.Unix(2, 0),
-			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(10)), twoDec.Add(OneSec.QuoInt64(10))),
+			record:    newRecord(poolId, time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			newTime:   time.Unix(2, 0),
+			expRecord: newExpRecord(oneDec.Add(OneSec.MulInt64(10)), twoDec.Add(OneSec.QuoInt64(10))),
 		},
 		"larger time interval": {
-			record:          newRecord(time.Unix(11, 0), sdk.NewDec(10), oneDec, twoDec),
-			interpolateTime: time.Unix(55, 0),
-			expRecord:       newExpRecord(oneDec.Add(OneSec.MulInt64(44*10)), twoDec.Add(OneSec.MulInt64(44).QuoInt64(10))),
+			record:    newRecord(poolId, time.Unix(11, 0), sdk.NewDec(10), oneDec, twoDec),
+			newTime:   time.Unix(55, 0),
+			expRecord: newExpRecord(oneDec.Add(OneSec.MulInt64(44*10)), twoDec.Add(OneSec.MulInt64(44).QuoInt64(10))),
 		},
-		"same time": {
-			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
-			interpolateTime: time.Unix(1, 0),
-			expRecord:       newExpRecord(oneDec, twoDec),
+		"same time, accumulator should not change": {
+			record:    newRecord(poolId, time.Unix(1, 0), sdk.NewDec(10), oneDec, twoDec),
+			newTime:   time.Unix(1, 0),
+			expRecord: newExpRecord(oneDec, twoDec),
 		},
 		// TODO: Overflow tests
```

Comment:
Is there an issue for this (i think there is)? If so, can we link to it here?

Reply:
There are comments all over twap module as of right now considering over flow.
https://github.com/osmosis-labs/osmosis/issues/2412